ludwig/datasets/sst2/sst_utils.py

```diff
@@ -0,0 +1,129 @@
+from ludwig.datasets.mixins.download import ZipDownloadMixin
+from ludwig.datasets.mixins.load import CSVLoadMixin
+from ludwig.datasets.mixins.process import *
+
+
+class SST(ABC, ZipDownloadMixin, MultifileJoinProcessMixin, CSVLoadMixin,
+          BaseDataset):
+    """The SST2 dataset.
+
+    This dataset is constructed using the Stanford Sentiment Treebank Dataset.
+    This dataset contains binary labels (positive or negative) for each sample.
+
+    The original dataset specified 5 labels:
+    very negative, negative, neutral, positive, very positive with
+    the following cutoffs:
+    [0, 0.2], (0.2, 0.4], (0.4, 0.6], (0.6, 0.8], (0.8, 1.0]
+
+    This class pulls in an array of mixins for different types of functionality
+    which belongs in the workflow for ingesting and transforming
+    training data into a destination dataframe that can be use by Ludwig.
+    """
+
+    def __init__(self, dataset_name, cache_dir=DEFAULT_CACHE_LOCATION):
+        super().__init__(dataset_name=dataset_name, cache_dir=cache_dir)
+
+    @staticmethod
+    @abstractmethod
+    def get_sentiment_label(id2sent, phrase_id):
+        pass
+
+    def process_downloaded_dataset(self):
+        sentences_df = pd.read_csv(
+            os.path.join(self.raw_dataset_path,
+                         'SST-2/original/datasetSentences.txt'),
+            sep=('\t'))
+        datasplit_df = pd.read_csv(
+            os.path.join(self.raw_dataset_path,
+                         'SST-2/original/datasetSplit.txt'),
+            sep=',')
+
+        phrase2id = {}
+        with open(os.path.join(self.raw_dataset_path,
+                               'SST-2/original/dictionary.txt')) as f:
+            Lines = f.readlines()
+            for line in Lines:
+                if line:
+                    split_line = line.split('|')
+                    phrase2id[split_line[0]] = int(split_line[1])
+
+        id2sent = {}
+        with open(os.path.join(self.raw_dataset_path,
+                               'SST-2/original/sentiment_labels.txt')) as f:
+            Lines = f.readlines()
+            for line in Lines:
+                if line:
+                    split_line = line.split('|')
+                    try:
+                        id2sent[int(split_line[0])] = float(split_line[1])
+                    except ValueError:
+                        pass
+
+        def format_sentence(sent):
+            formatted_sent = ' '.join(
+                [w.encode('latin1').decode('utf-8')
+                 for w in sent.strip().split(' ')])
+            formatted_sent = formatted_sent.replace('-LRB-', '(')
+            formatted_sent = formatted_sent.replace('-RRB-', ')')
+            return formatted_sent
+
+        def get_sentence_idxs(split):
+            return set(
+                datasplit_df[
+                    datasplit_df['splitset_label'] == split
+                    ]['sentence_index']
+            )
+
+        def get_sentences(sentences_idxs):
```

Comment:
could you add some contextual comments to this method mostly so that someone has a big picture on what this code is trying to accomplish before reading any code

Reply:
Good idea, @ANarayan maybe you could add come comments for all those internal functions to clarify what they are doing?

Reply:
Yes @w4nderlust, I will add comments!